girder/api/v1/resource.py

```diff
@@ -139,6 +140,124 @@ def _getResourceModel(self, kind, funcName=None):
             raise RestException('Invalid resources format.')
         return model
 
+    def _lookupToken(self, token, parentId):
+        """
+        Find a particular child resource by name or throw an exception.
+        :param token: the name of the child resource to find
+        :param parentId: id of the parent to search
+        :returns: the child resource.
+        """
+
+        filterObject = {'name': token}
+        candidateParent = None
+
+        # figure out the parentId's type
+        candidateModelNames = ('collection', 'user', 'folder', 'item')
+        for modelName in candidateModelNames:
+            candidateParent = self.model(modelName).findOne({'_id': parentId})
+
+            if candidateParent is not None:
+                if modelName in ('collection', 'user'):
+                    filterObject.update(
+                        parentId=parentId,
+                        baseParentId=parentId,
+                        baseParentType=modelName)
+                else:
+                    filterObject.update(
+                        parentId=parentId,
+                        baseParentId=candidateParent['baseParentId'],
+                        baseParentType=candidateParent['baseParentType'])
+                break
+
+        if candidateParent is None:
+            raise RestException(
+                'Parent resource not found: {}'.format(parentId))
+
+        candidateChild = self.model('folder').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'folder'
+
+        filterObject['folderId'] = filterObject.pop('parentId')
+        del filterObject['baseParentId']
+        del filterObject['baseParentType']
+        candidateChild = self.model('item').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'item'
+
+        filterObject['itemId'] = filterObject.pop('folderId')
+        candidateChild = self.model('file').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'file'
+
+        raise RestException('Child resource not found: {}->{}'.format(
+            parentId, token))
+
+    def _lookupPath(self, path, user=None):
+        """
+        Find a particular resource by path or throw an exception.  The given
+        user must have read access to all resources featured in the path.
+        :param path: the path of the object to find
+        :param user: the user to search as
+        :returns: the resource.
+        """
+
+        if not path.startswith('/'):
+            raise RestException('Invalid path format')
+
+        pathArray = path[1:].split('/')
```

Comment:
I vote we make the leading slash optional since we are always going to be getting an "absolute" path.  So we could change this to

    pathArray = [t for t in path.split('/') if t]